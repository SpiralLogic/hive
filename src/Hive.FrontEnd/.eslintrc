{
  "plugins": [
    "@typescript-eslint",
    "react-hooks",
    "testing-library",
    "jest-dom",
    "unicorn"
  ],
  "env": {
    "browser": true,
    "es6": true
  },
  "extends": [
    "kentcdodds",
    "plugin:jest/recommended",
    "plugin:jest-dom/recommended",
    "plugin:prettier/recommended",
    "plugin:testing-library/react",
    "plugin:testing-library/dom",
    "prettier",
    "plugin:unicorn/recommended",
    "preact"
  ],
  "rules": {
    "unicorn/no-null": "off",
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": [
      "error"
    ]
  },
  "globals": {
    "Atomics": "readonly",
    "SharedArrayBuffer": "readonly",
    "h": true
  },
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "./tsconfig.json",
    "jsx": true
  },
  "overrides": [
    {
      "files": [
        "**/*.jsx",
        "**/*.tsx"
      ],
      "rules": {
        "unicorn/filename-case": "off"
      }
    }
  ],
  "settings": {
    "react": {
      "pragma": "h"
    },
    "propWrapperFunctions": [
      "forbidExtraProps",
      {
        "property": "freeze",
        "object": "Object"
      }
    ],
    "linkComponents": [
      "Hyperlink",
      {
        "name": "Link",
        "linkAttribute": "to"
      }
    ]
  },
  "ignorePatterns": [
    "/node_modules/*/**",
    "/*.config.js",
    "/src/public/*/**",
    "**/__coverage__/**",
    "/src/index.tsx",
    "/src/index.html",
    "/build.js",
    "/inject.shim.js"
  ]
}
