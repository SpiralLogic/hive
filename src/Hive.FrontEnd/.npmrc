node-version=20.11.1