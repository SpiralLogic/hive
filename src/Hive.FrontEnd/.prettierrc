{
  "singleQuote": true,
  "printWidth": 110,
  "proseWrap": "always",
  "bracketSameLine": true,
  "endOfLine": "auto",
  "trailingComma": "es5",
  "tabWidth": 2
}
