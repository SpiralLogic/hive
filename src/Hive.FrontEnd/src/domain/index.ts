export * from './cell';
export * from './game-state';
export * from './hex-coordinate';
export * from './move';
export * from './player';
export * from './row';
export * from './tile';
export * from './creature';
