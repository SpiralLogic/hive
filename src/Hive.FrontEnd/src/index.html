<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="description" content="Hive online board game">
  <title>Hive</title>
  [css]
  [scripts]    
</head>
</html>
