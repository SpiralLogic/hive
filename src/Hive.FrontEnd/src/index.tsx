import { h, render } from 'preact';
import App from './components/App';
import { serverConnectionFactory } from './services';
import creatures from './svg/creatures.svg?raw';

import GameEngine from './services/game-engine';

document.body.insertAdjacentHTML('beforeend', `${creatures}`);
document.title = import.meta.env.HIVE_APP_TITLE;

const [, , gameId, currentPlayer] = window.location.pathname.split('/');
const engine = new GameEngine({ gameId, currentPlayer });
const properties = {
  engine,
  connectionFactory: serverConnectionFactory,
};

render(h(App, properties), document.body);

export { removeOtherPlayerMoves } from './utilities/hextille';
export { cellKey } from './utilities/hextille';
console.log(import.meta.env);
