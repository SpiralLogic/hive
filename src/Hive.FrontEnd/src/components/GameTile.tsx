import { FunctionComponent } from 'preact';
import { useEffect, useState } from 'preact/hooks';

import { PlayerId, Tile as TileType } from '../domain';
import { TileAction } from '../services';
import { useClassReducer } from '../utilities/class-reducer';
import { useHiveDispatchListener, dispatchHiveEvent } from '../utilities/dispatcher';
import { handleDrop, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import Tile from './Tile';

const tileSelector = `[tabindex].tile`;
const cellSelector = `[tabindex][role="cell"]`;
type Properties = TileType & { stacked?: boolean; currentPlayer: PlayerId };

const handleMouseLeave = (event: { currentTarget: HTMLElement }) => event.currentTarget.blur();

const GameTile: FunctionComponent<Properties> = (properties) => {
  const { currentPlayer, stacked, ...tile } = properties;
  const { id, moves, creature, playerId } = tile;
  const [focus, setFocus] = useState(tileSelector);
  const [classes, setClassList] = useClassReducer(`player${playerId} hex`);
  if (stacked) setClassList({ type: 'add', classes: ['stacked'] });

  const deselect = (fromEvent = false) => {
    if (!classes.includes('selected')) return;
    setClassList({ type: 'remove', classes: ['selected'] });
    dispatchHiveEvent({ type: 'tileDeselected', tile, fromEvent });
  };

  const select = (fromEvent = false) => {
    if (classes.includes('selected')) return;
    dispatchHiveEvent({ type: 'tileClear' });
    setClassList({ type: 'add', classes: ['selected'] });
    if (currentPlayer != playerId) return;
    dispatchHiveEvent({ type: 'tileSelected', tile, fromEvent });
  };

  useHiveDispatchListener<TileAction>('tileSelect', (event: TileAction) => {
    if (event.tile.id === id) select(true);
  });

  useHiveDispatchListener<TileAction>('tileDeselect', (event: TileAction) => {
    if (event.tile.id === id) deselect(true);
  });

  useHiveDispatchListener('tileClear', () => {
    deselect(true);
    setFocus('');
  });

  const handleDragStart = (event: DragEvent) => {
    event.stopPropagation();
    select();
    setClassList({ type: 'add', classes: ['before-drag'] });
    setTimeout(() => setClassList({ type: 'remove', classes: ['before-drag'] }), 1);
  };

  const handleDragEnd = () => {
    dispatchHiveEvent({ type: 'tileClear' });
    dispatchHiveEvent({ type: 'tileDropped', tile });
  };

  const handleClick = (event: MouseEvent) => {
    event.stopPropagation();
    return classes.includes('selected') ? deselect() : select();
  };

  const handleKeyDown = (event: KeyboardEvent) => {
    if (handleKeyboardNav(event) || !isEnterOrSpace(event)) return;
    event.stopPropagation();
    if (classes.includes('selected')) {
      deselect();
    } else {
      select();
      setFocus(cellSelector);
    }
  };
  useEffect(() => {
    if (!focus) return;
    const focusElement = document.querySelector<HTMLElement>(focus);
    focusElement?.focus();
    setFocus('');
  }, [focus]);

  const attributes = {
    title: `Player-${playerId} ${creature}`,
    class: classes,
    draggable: moves.length > 0 ? true : undefined,
    tabindex: moves.length > 0 ? 0 : undefined,
    creature,
  };

  const handlers =
    moves.length > 0
      ? {
          onclick: handleClick,
          ondragstart: handleDragStart,
          ondragend: handleDragEnd,
          onkeydown: handleKeyDown,
          onmouseleave: handleMouseLeave,
          ondrop: handleDrop,
        }
      : {
          ondrop: handleDrop,
        };
  return <Tile {...attributes} {...handlers} />;
};
GameTile.displayName = 'GameTile';
export default GameTile;
