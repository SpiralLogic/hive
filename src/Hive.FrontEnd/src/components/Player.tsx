import '../css/player.css';

import { FunctionComponent, toChildArray } from 'preact';
import { useEffect, useState } from 'preact/hooks';

import { PlayerId } from '../domain';
import { useClassReducer } from '../utilities/class-reducer';

type Properties = {
  name: string;
  id: PlayerId;
};
const Player: FunctionComponent<Properties> = (properties) => {
  const { name, id, children } = properties;
  const [classes, setClassList] = useClassReducer(`player player${id}`);
  const [hidden, setHidden] = useState(false);
  const hasChildren = toChildArray(children).length === 0;

  useEffect(() => {
    if (hasChildren) {
      setTimeout(() => setClassList({ type: 'add', classes: ['hide'] }), 100);
    }
  }, [hasChildren, setClassList]);

  const ontransitionend = () => {
    setHidden(hasChildren);
  };

  const handlers = { onTransitionEnd: ontransitionend, onAnimationEnd: ontransitionend };

  return hidden ? null : (
    <section class={classes} aria-label={`${name}'s unplaced pieces`} {...handlers}>
      <h2>{name}</h2>
      <div class="tiles">{children}</div>
    </section>
  );
};

Player.displayName = 'Player Tiles';
export default Player;
