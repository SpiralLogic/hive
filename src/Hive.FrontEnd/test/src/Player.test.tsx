import { fireEvent, render, screen } from '@testing-library/preact';
import Player from '../../src/components/Player';
import GameTile from '../../src/components/GameTile';
import { HexCoordinates } from '../../src/domain';

const setUp = () => {
  global.window.history.replaceState({}, global.document.title, `/game/33/1`);
  const playerProperties = { id: 1, name: 'Player 1' };
  return render(
    <Player {...playerProperties}>
      <GameTile
        id={1}
        currentPlayer={1}
        playerId={1}
        creature={'thing'}
        moves={new Array<HexCoordinates>()}
      />
    </Player>
  );
};

describe('<Player>', () => {
  it('has rendered with player name', () => {
    setUp();
    expect(screen.getByRole('heading')).toHaveTextContent(/Player 1/);
  });

  it('player is rendered with their tiles', () => {
    setUp();
    expect(screen.getByTitle(/Player-1/)).toBeInTheDocument();
  });

  it(`player is hidden when last tile is played`, async () => {
    vi.useFakeTimers();

    const { rerender } = setUp();

    expect(screen.getByLabelText(/player 1/i)).not.toHaveClass('hide');

    rerender(<Player id={1} name="Player 1" />);
    vi.runAllTimers();

    expect(await screen.findByLabelText(/player 1/i)).toHaveClass('hide');

    fireEvent.transitionEnd(screen.getByLabelText(/player 1/i));

    expect(screen.queryByLabelText(/player 1/i)).toHaveClass('hidden');
  });

  it('renders', () => {
    expect(setUp()).toMatchSnapshot();
  });
});
