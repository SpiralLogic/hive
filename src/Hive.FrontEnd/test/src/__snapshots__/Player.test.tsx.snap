// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Player> renders 1`] = `
<section
  aria-label="Player 1's unplaced pieces"
  class="player player1"
>
  <h2>
    Player 1
  </h2>
  <div
    class="tiles"
  >
    <div
      class="player1 hex tile thing"
      title="Player-1 thing"
    >
      <svg
        viewBox="0 0 100 100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          href="#hex"
        />
        <use
          class="creature"
          href="#thing"
        />
      </svg>
    </div>
  </div>
</section>
`;
